integration_test.go

```diff
@@ -3,51 +3,139 @@ package sockchat
+				ws.Write(t, NewSocketMessage(LoginAction, LoginRequest{Nick: ValidUserNick, Password: ValidUserPassword}))
+				ws.AssertEventReceivedWithin(t, "logged_in:"+ValidUserNick, 100*time.Millisecond)
+				logged_in_conns <- ws
 			}(conn)
+			<-logged_in_conns
+		}
+
+	})
+
+	t.Run("Both connections receive information about joining the channel when one of the connections sent request to join", func(t *testing.T) {
+		conns[0].Write(t, NewSocketMessage(JoinAction, ChannelRequest{"foo"}))
+		for _, conn := range conns {
+			conn.AssertEventReceivedWithin(t, UserJoinedChannelEvent, 100*time.Millisecond)
 		}
+	})
 
+	t.Run("Second user's conn receives event about new message in the channel they're in", func(t *testing.T) {
+		msgText := "Baz"
+		conns[1].Write(t, NewSocketMessage(SendMessageAction, SendMessageRequest{Channel: "foo", Text: msgText}))
+		select {
+		case message_received := <-conns[0].MessageStash:
+			assert.Equal(t, msgText, UnmarshalMessageEvent(message_received.Payload).Text)
+			return
+		case <-time.After(100 * time.Millisecond):
+			t.Error("Message not received by second connection")
+		}
 	})
+
+	t.Run("Both connections get notified when user leaves the channel", func(t *testing.T) {
+		conns[0].Write(t, NewSocketMessage(LeaveAction, ChannelRequest{"foo"}))
+		for _, conn := range conns {
+			conn.AssertEventReceivedWithin(t, YouLeftChannelEvent, 100*time.Millisecond)
+		}
+	})
+
+	t.Run("After leaving the channel, no connection can send messages to it", func(t *testing.T) {
+		conns[1].Write(t, NewSocketMessage(SendMessageAction, SendMessageRequest{Channel: "foo", Text: "Baz"}))
+		conns[1].AssertEventReceivedWithin(t, InvalidRequestEvent, 100*time.Millisecond)
+	})
+}
+
+func TestMultipleUsersSync(t *testing.T) {
+
+	store, _ := NewChannelStore()
+	store.CreateChannel("foo")
+	users := &userStoreDouble{}
+	server := httptest.NewServer(NewSockChatServer(store, users))
+	wsURL := GetWsURL(server.URL)
+
+	conns := make([]*TestWS, 2)
+	for i := 0; i < 2; i++ {
+		conns[i] = NewTestWS(t, wsURL)
+		defer conns[i].Close()
+	}
+	defer server.Close()
+
+	t.Run("Two users can log in using two connections at the same time", func(t *testing.T) {
+		done := make(chan bool, 2)
+		go func() {
+			conns[0].Write(t, NewSocketMessage(LoginAction, LoginRequest{Nick: ValidUserNick, Password: ValidUserPassword}))
+			conns[0].AssertEventReceivedWithin(t, "logged_in:"+ValidUserNick, 100*time.Millisecond)
+			done <- true
+		}()
+		go func() {
+			conns[1].Write(t, NewSocketMessage(LoginAction, LoginRequest{Nick: ValidUser2Nick, Password: ValidUserPassword}))
+			conns[1].AssertEventReceivedWithin(t, "logged_in:"+ValidUser2Nick, 100*time.Millisecond)
+			done <- true
+		}()
+		for i := 0; i < 2; i++ {
+			<-done
+		}
+	})
+
+	t.Run("Two users can join the same channel", func(t *testing.T) {
+		go conns[0].Write(t, NewSocketMessage(JoinAction, ChannelRequest{"foo"}))
+		go conns[1].Write(t, NewSocketMessage(JoinAction, ChannelRequest{"foo"}))
```

Comment:
`TestWS` is a client using `gorilla/websocket` library so it should protect concurrent writes for example with a mutex. Current test scenarios show races caused by `conns[1].Write` in 2 different goroutines from 2 different test cases:
- goroutine spawned in `"Two users can join the same channel"`
- goroutine running the test `"One of the users disconnects - check if the other user gets notified"`